core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/Histogram.java

```diff
@@ -40,46 +48,56 @@
      * @return  The buckets of this histogram (each bucket representing an interval in the histogram)
      */
     @Override
-    List<? extends Bucket> getBuckets();
+    List<Bucket> getBuckets();
 
 
     /**
      * A strategy defining the order in which the buckets in this histogram are ordered.
      */
     abstract class Order implements ToXContent {
 
-        public static final Order KEY_ASC = new InternalOrder((byte) 1, "_key", true, new Comparator<InternalHistogram.Bucket>() {
+        private static int compareKey(Histogram.Bucket b1, Histogram.Bucket b2) {
+            if (b1 instanceof InternalHistogram.Bucket) {
+                return Double.compare(((InternalHistogram.Bucket) b1).key, ((InternalHistogram.Bucket) b2).key);
+            } else if (b1 instanceof InternalDateHistogram.Bucket) {
+                return Long.compare(((InternalDateHistogram.Bucket) b1).key, ((InternalDateHistogram.Bucket) b2).key);
+            } else {
+                throw new AssertionError("Unexpected impl: " + b1.getClass());
```

Comment:
Can you throw something else? It just makes me uncomfortable to throw AssertionError.